pkg/routing/signal.go

```diff
@@ -94,49 +93,248 @@ func (r *signalClient) StartParticipantSignal(
 
 	go func() {
 		r.active.Inc()
 		defer r.active.Dec()
 
-		var err error
-		for msg := range stream.Channel() {
-			if err = resChan.WriteMessage(msg.Response); err != nil {
-				break
+		err = CopySignalStreamToMessageChannel[*rpc.RelaySignalRequest, *rpc.RelaySignalResponse](
+			stream,
+			resChan,
+			signalResponseMessageReader{},
+			r.config,
+		)
+		l.Debugw("participant signal stream closed", "error", err)
+
+		resChan.Close()
+	}()
+
+	return connectionID, sink, resChan, nil
+}
+
+type signalRequestMessageWriter struct{}
+
+func (e signalRequestMessageWriter) WriteOne(seq uint64, msg proto.Message) *rpc.RelaySignalRequest {
+	return &rpc.RelaySignalRequest{
+		Seq:     seq,
+		Request: msg.(*livekit.SignalRequest),
+	}
+}
+
+func (e signalRequestMessageWriter) WriteMany(seq uint64, msgs []proto.Message) *rpc.RelaySignalRequest {
+	r := &rpc.RelaySignalRequest{
+		Seq:      seq,
+		Requests: make([]*livekit.SignalRequest, 0, len(msgs)),
+	}
+	for _, m := range msgs {
+		r.Requests = append(r.Requests, m.(*livekit.SignalRequest))
+	}
+	return r
+}
+
+type signalResponseMessageReader struct{}
+
+func (e signalResponseMessageReader) Read(rm *rpc.RelaySignalResponse) ([]proto.Message, error) {
+	msgs := make([]proto.Message, 0, len(rm.Responses)+1)
+	if rm.Response != nil {
+		msgs = append(msgs, rm.Response)
+	}
+	for _, m := range rm.Responses {
+		msgs = append(msgs, m)
+	}
+	return msgs, nil
+}
+
+type RelaySignalMessage interface {
+	proto.Message
+	GetSeq() uint64
+}
+
+type SignalMessageWriter[SendType RelaySignalMessage] interface {
+	WriteOne(seq uint64, msg proto.Message) SendType
+	WriteMany(seq uint64, msgs []proto.Message) SendType
+}
+
+type SignalMessageReader[RecvType RelaySignalMessage] interface {
+	Read(msg RecvType) ([]proto.Message, error)
+}
+
+func CopySignalStreamToMessageChannel[SendType, RecvType RelaySignalMessage](
+	stream psrpc.Stream[SendType, RecvType],
+	ch *MessageChannel,
+	reader SignalMessageReader[RecvType],
+	config config.SignalRelayConfig,
+) error {
+	r := &signalMessageReader[SendType, RecvType]{
+		reader: reader,
+		config: config,
+	}
+	for msg := range stream.Channel() {
+		var res []proto.Message
+		res, err := r.Read(msg)
+		if err != nil {
+			return err
+		}
+		for _, r := range res {
+			if err = ch.WriteMessage(r); err != nil {
+				return err
 			}
-			for _, res := range msg.Responses {
-				if err = resChan.WriteMessage(res); err != nil {
-					break
-				}
+		}
+	}
+	return stream.Err()
+}
+
+type signalMessageReader[SendType, RecvType RelaySignalMessage] struct {
+	seq    uint64
+	reader SignalMessageReader[RecvType]
+	config config.SignalRelayConfig
+}
+
+func (r *signalMessageReader[SendType, RecvType]) Read(msg RecvType) ([]proto.Message, error) {
+	res, err := r.reader.Read(msg)
+	if err != nil {
+		return nil, err
+	}
+
+	if r.config.MinVersion >= 1 {
+		if r.seq < msg.GetSeq() {
+			return nil, errors.New("participant signal message dropped")
```

Comment:
Just for my own understanding, guessing this is controller reading from media node (signal response) and a gap is treated as error and the behaviour corresponds to this in your PR commit notes
```
if there is a dropped message close the stream and restart the client
```
Would that work always? Let us the client (by client I mean the end user client) sends `AddTrack`. It would expect a response. But, if that response is dropped between relay signal client (controller) and relay signal server (media node), restarting the relay signal client may not be enough? Would restarting be able to get that signal response and send it to the edge client?

Reply:
I think returning an error here will cause the websocket to close (when resChan is closed). 

this should kick client into a reconnection sequence.

Reply:
this is correct. failing here closes the stream and the user reconnects.